research/conditional/utils/model_utils.py

```diff
@@ -49,90 +48,96 @@
 from research.conditional.moe_layers.expert_choice import ExpertChoiceFF
 from research.conditional.moe_layers.token_choice import TokenChoiceFF
 from research.conditional.moe_layers.ff_timed import FeedForwardTimed
+from torch.nn.parallel import DistributedDataParallel as DDP
 
 
-def make_loss_function(loss_checkpoint_chungs: int):
+def make_loss_and_backprop_function(loss_checkpoint_chungs: int):
     if loss_checkpoint_chungs == 0:
-        return calculate_llm_loss
+        return calculate_llm_loss_and_backward_pass
     else:
-        return partial(chungized_llm_loss, n_chungs=loss_checkpoint_chungs)
+        return partial(
+            chungized_llm_loss_and_backward_pass, n_chungs=loss_checkpoint_chungs
+        )
 
 
-def chungized_llm_loss(
+def chungized_llm_loss_and_backward_pass(
     batch: LLMBatch,
     model: torch.nn.Module,
     mixed_precision: bool,
     vocab_size: int,
+    scaler: torch.cuda.amp.GradScaler,
     n_chungs: int,
+    backward_pass: bool,
+    num_accumulated_batches: int,
 ):
+    if isinstance(model, DDP):
+        model = model.module
     input_tokens = batch.input_ids
     gt_tokens = batch.target_ids
     mask = batch.should_calculate_loss
-
-    def make_custom_forward():
-        def custom_forward(*inputs):
-            output = model.head(inputs[0])
-            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
-                gt = inputs[1]
-                mask = inputs[2]
-                gt = gt.to(output.device)
-                loss = F.cross_entropy(
-                    output.reshape(-1, vocab_size),
-                    gt.reshape(-1).long(),
-                    reduction="none",
-                )
-
-                correct_tokens = gt.long() == output.argmax(dim=-1)
-                correct_tokens = correct_tokens.long().reshape(-1) * mask.reshape(-1)
-                correct_tokens = correct_tokens.sum()
-
-                total_tokens = mask.sum()
-
-            return loss[mask.reshape(-1) == 1], correct_tokens, total_tokens
-
-        return custom_forward
+    num_masked_tokens = mask.sum()
 
     with torch.autocast(
         device_type="cuda", enabled=mixed_precision, dtype=torch.float16
     ):
-        encoder_output = model.encoder(input_tokens)
-        chunged_inputs = torch.chunk(encoder_output, n_chungs, dim=0)
+        encoder_output: torch.Tensor = model.encoder(input_tokens)
+        encoder_output_det = encoder_output.detach()
+        if backward_pass:
+            encoder_output_det.requires_grad = True
```

Comment:
this detachment is due to the fact that you calculate loss manually, right? We couldn't leave the output attached?

Reply:
Yes, if we leave it attached the gradient would pass further into the network and we don't want to do that for performance reasons

Reply:
I see, thanks!